src/control/server/harness.go

```diff
@@ -256,20 +254,96 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
 		}
 	}
 
-	// now monitor them
-	select {
-	case <-parent.Done():
-		return parent.Err()
-	case err := <-errChan:
-		// If we receive an error from any instance, shut them all down.
-		// TODO: Restart failed instances rather than shutting everything
-		// down.
-		h.log.Errorf("instance error: %s", err)
-		if err != nil {
-			return errors.Wrap(err, "Instance error")
+	return nil
+}
+
+// monitorInstances listens for exit results from instances or harness and will
+// return only when all harness instances are stopped and restartInstances
+// signal is received.
+func (h *IOServerHarness) monitorInstances(ctx context.Context) error {
+	h.log.Debug("monitoring instances")
+	for {
+		select {
+		case <-ctx.Done(): // received when harness is exiting
+			h.log.Debug("harness exiting")
+			return ctx.Err()
+		case err := <-h.errChan: // received when instance exits
+			// TODO: Restart failed instances on unexpected exit.
+			allInstancesStopped := !h.HasStartedInstances()
+			msg := fmt.Sprintf("instance exited: %v", err)
+			if allInstancesStopped {
+				msg += ", all instances stopped!"
+			}
+			h.log.Info(msg)
+		case <-h.restartInstances: // trigger harness to restart instances
+			h.log.Debug("restart instances")
+			if h.HasStartedInstances() {
+				return errors.New("cannot restart when instances are running")
+			}
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// Start starts all configured instances, waits for them to be ready and then
+// loops monitoring instance exit, harness exit and harness restart signals.
+func (h *IOServerHarness) Start(parent context.Context, membership *system.Membership) error {
+	if h.IsStarted() {
+		return errors.New("can't start: harness already started")
+	}
+
+	// Now we want to block any RPCs that might try to mess with storage
+	// (format, firmware update, etc) before attempting to start I/O servers
+	// which are using the storage.
+	h.Lock()
+	h.started = true
+	h.Unlock()
+
+	ctx, shutdown := context.WithCancel(parent)
+	defer shutdown()
+
+	for {
+		if err := h.startInstances(ctx, membership); err != nil {
+			return err
+		}
+		if err := h.waitInstancesReady(ctx); err != nil {
+			return err
+		}
+		if err := h.monitorInstances(ctx); err != nil {
```

Comment:
I think it'd be worth adding a comment to explain the logic here... I understand it now but I had to study the code. In particular, it's not obvious that sending a restart signal after all instances have stopped results in another loop iteration.

Reply:
have added comments